prevent-broker-scale-down-if-contains-partition .md

```diff
@@ -0,0 +1,63 @@
+If the partitions are still available on the broker which is going to be scaled, then we should get some warning in the logs which will prompt users to do the reassignment and hopefully reduce the chance of data loss.
+
+## Proposal
+
+This proposal suggest how we can add the check to detect if the broker still contains any partitions and what to do if the broker scale down is not possible.
+
+### Checking if the broker contains partition
+The best way to detect if the broker contains any partitions is to query the topic metadata.
+To do this, we will have to make use of the `AdminClient` class.
+We should probably put the logic for this check inside the `scaleDown` method present in the `KafkaReconciler` class.
+The `scaleDown` method gets triggered when we scale down the no. of replicas in the `Kafka CR`.
+So this check will target the broker which are going to be removed i.e. the pods with the highest number.
+
+Methods to be created in the process:
+
+- `getTopicNamesAndDescriptions` (Boolean Type) - This method will be placed inside the `scaleDown()` method in `KafkaReconciler` class and will provide us with the Collection of `TopicDescription`.
+- `describeTopics()` and `topicNames()`  - These two methods will provide us with the description and names of all the topics ie first we will get the names of the topics using the `topicNames` method and using these names we will extract the description of the topics.
+- `brokerContainsPartitions` - This method will take as input a `Collection<TopicDescription>` and then generate a map of type `HashMap<Integer podId, Boolean containsPartitions>`.
+
+### Flow :
+
+When the replicas are changed inside the Kafka resource, reconciliation will happen.
+During this reconciliation, we move to the `scaleDown()` method present in the `KafkaReconciler` class.
+The `scaleDown()` method contains the complete mechanism on how the broker is scaled down. So there, we will first create our `adminClient` and provide it with the Secrets.
+Then we will call the `getTopicNamesAndDescriptions()` method which will give us the `Collection<TopicDescription>`. 
+
+The `getTopicNamesAndDescriptions()` method will call the `topicNames()` method to get the set of topic names and then these topic names will be used by the `describeTopics()` methods to get the `Collection<TopicDescription>`. 
+Once the `Collection<TopicDescription>` is computed we will move back to the `scaledown()` method and there we can call `brokerContainsPartition()` method which will take the `Collection<TopicDescription>` and then create a `hasAssignedTopicPartitions` map which contains key as `podId` and value as `boolean` type.
+Now we can use this map to check if the broker contains any partitions or not.
+
+### What to do if a broker contains partitions?
+If a partition is detected over a broker then we should trigger some log warnings and let the partitions remain the same as they were before.
```

Comment:
Again, fail the reconciliation, since logging won't actually prevent data loss.

Reply:
But updating the Kafka CR with a warning condition.